src/psyclone/tests/gocean1p0_opencl_test.py

```diff
@@ -176,3 +178,128 @@ def test_set_arg_const_scalar():
         otrans.apply(sched)
     assert ("Cannot generate OpenCL for Invokes that contain kernels with "
             "arguments passed by value" in str(err))
+
+
+def test_opencl_kernel_code_generation():
+    ''' Tests that gen_ocl method of the GOcean Kernel Schedule generates
+    the expected OpenCL code.
+    '''
+
+    psy, _ = get_invoke("single_invoke.f90", API, idx=0)
+    sched = psy.invokes.invoke_list[0].schedule
+    kernel = sched.children[0].children[0].children[0]  # compute_cu kernel
+    kschedule = kernel.get_kernel_schedule()
+
+    expected_code = (
+        "__kernel void compute_cu_code(\n"
+        "    __global double * restrict cu,\n"
+        "    __global double * restrict p,\n"
+        "    __global double * restrict u\n"
+        "    ){\n"
+        "    int cuLEN2 = get_global_size(1);\n"
+        "    int pLEN2 = get_global_size(1);\n"
+        "    int uLEN2 = get_global_size(1);\n"
+        "    int i = get_global_id(0);\n"
+        "    int j = get_global_id(1);\n"
+        )
+
+    assert expected_code in kschedule.gen_ocl()
+
+    # TODO: At the moment, due to fparser/171, the body of compute_cu
+    # is not generated and can not be included on the expected code,
+    # this cause a line without coverage in the tests.
+
+
+def test_opencl_kernel_gen_different_datatypes():
+    ''' Tests that gen_ocl method of the GOcean Kernel Schedule generates
+    the expected OpenCL argument/variable declarations.
+    '''
+    kschedule = GOKernelSchedule('test')
+    kschedule.symbol_table.declare("i", "integer", [])
+    kschedule.symbol_table.declare("j", "integer", [])
+    kschedule.symbol_table.declare("intarg", "integer", [])
+    kschedule.symbol_table.declare("realarg", "real", [])
+    kschedule.symbol_table.declare("chararg", "character", [])
+    kschedule.symbol_table.declare("arrayarg", "real", [3, 4, 5, 3])
+    kschedule.symbol_table.declare("intvar", "integer", [])
+    kschedule.symbol_table.declare("realvar", "real", [])
+    kschedule.symbol_table.declare("charvar", "character", [])
+    kschedule.symbol_table.declare("arrayvar", "real", [3, 4, 5, 3])
+    kschedule.symbol_table.specify_argument_list(["i", "j", "intarg",
+                                                  "realarg", "chararg",
+                                                  "arrayarg"])
+    opencl = kschedule.gen_ocl()
+
+    # Check Arguments are part (or not) of the generated opencl
+    assert "int i," not in opencl
+    assert "int j," not in opencl
+    assert "int intarg" in opencl
+    assert "double realarg" in opencl
+    assert "char chararg" in opencl
+    assert "__global double * restrict arrayarg" in opencl
+
+    # Check local variables are declared
+    assert "int intvar;" in opencl
+    assert "double realvar;" in opencl
+    assert "char charvar;" in opencl
+    assert "double realvar;" in opencl
+    assert "double * arrayvar;" in opencl
+    assert "int arrayargLEN1 = get_global_size(0);" not in opencl
```

Comment:
This was wrong, I meant to left out the last dimension as is not needed on the array index expressions, but in retrospective I decided to include it anyway as it may appear on other places (loop bounds?) and if it does not appear, it is easy for the compiler to eliminate.